app/models/concepts_graph.rb

```diff
@@ -0,0 +1,98 @@
+# ConceptGraph transforms a track configuration to a form compatible for manipulating for
+# displaying the progression of the track
+class ConceptsGraph
+  attr_reader(
+    :language,
+    :nodes,
+    :edges,
+    :node_lookup_by_concept,
+    :node_lookup_by_index,
+    :adjacency,
+    :has_cycle
+  )
+
+  def initialize(track_config)
+    @language = track_config['language']
+    @nodes = make_nodes(track_config['exercises'])
+    @node_lookup_by_concept = make_concept_lookup
+    @node_lookup_by_index = make_index_lookup
+    @edges = make_edges
+    @adjacency = GraphUtils.make_adjacency_list(self)
+    @has_cycle = GraphUtils.contains_cycle?(adjacency)
+  end
+
+  # Layout computes the ordering of the exercises in a track by creating
+  # a graph then traversing it to discover and memoize the progression
+  def to_layout
+    return [] if nodes.empty?
+    return :error if has_cycle
```

Comment:
Let's inline this. We only use the variable in one place so we don't need it extracted. This way we also can do it JIT.

```suggestion
    return :error if GraphUtils.contains_cycle?(adjacency)
```